src/tests/Microsoft.PowerFx.Core.Tests/DependencyFinderTests.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Linq;
+using Microsoft.PowerFx.Core.Syntax;
+using Microsoft.PowerFx.Core.Syntax.Visitors;
+using Xunit;
+
+namespace Microsoft.PowerFx.Core.Tests
+{
+    public class DependencyFinderTests
+    {
+        [Theory]
+        // simple cases
+        [InlineData("", "")]
+        [InlineData("5", "")]
+        [InlineData("true", "")]
+        [InlineData("\"hello\" & \"world\"", "")]
+        [InlineData("A", "A")]
+        [InlineData("Blank()", "")]
+        [InlineData("5 B", "")] // error
+        [InlineData("A As B", "")] // error
+        // Parent, Self, replaceable nodes not currently supported
+        [InlineData("Parent", "")]
+        [InlineData("Parent.A", "")]
+        [InlineData("Self", "")]
+        [InlineData("Self.A", "")]
+        [InlineData("%placeholder%", "")]
+        [InlineData("%placeholder%.A", "")]
+        // composite expressions
+        [InlineData("A;B;C", "A,B,C")]
+        [InlineData("[A,B,C]", "A,B,C")]
+        [InlineData("A.B", "A.B")]
+        [InlineData("A.B.C", "A.B.C")]
+        [InlineData("If(A)", "")] // error
+        [InlineData("If(A, B, C)", "A,B,C")]
+        [InlineData("If(A, B, C, D, E).P", "A,B.P,C,D.P,E.P")]
+        [InlineData("If(A, B, C).Name", "A,B.Name,C.Name")]
+        [InlineData("[If(A, B, C).Name,D,E]", "A,B.Name,C.Name,D,E")]
+        [InlineData("Switch(A)", "")] // Error
+        [InlineData("Switch(A, B)", "")] // Error
+        [InlineData("Switch(Slider1.Value, 20, Result1, 10, Result2, 0, Result3)", "Result1,Result2,Result3,Slider1.Value")]
+        [InlineData("Switch(Slider1.Value, 20, Result1, 10, Result2, 0, Result3).Name", "Result1.Name,Result2.Name,Result3.Name,Slider1.Value")]
+        [InlineData("Switch(Slider1.Value, 20, Result1, 10, Result2, 0, Result3, DefaultResult)", "Result1,Result2,Result3,DefaultResult,Slider1.Value")]
+        [InlineData("Switch(Slider1.Value, 20, Result1, 10, Result2, 0, Result3, DefaultResult).Name", "Result1.Name,Result2.Name,Result3.Name,DefaultResult.Name,Slider1.Value")]
+        [InlineData("With(A).Name", "")] // error
+        [InlineData("With({ B: 1, C: B.Inner } As Y, Y.C)", "B.Inner")]
+        [InlineData("With(A, B).Name", "A.B.Name,B.Name")]
+        [InlineData("With(A As B, B).Name", "A.Name")]
+        [InlineData("With(User, Age + 1)", "User.Age,Age")]
```

Comment:
Returning multiple paths isn't enough. You need a way of eventually converging on a coherent worldview.

For example, given a (potentially empty) list of input symbols with known types, you get back one set of dependencies for a given input formula. Some of the output symbols might be ambiguous or inferred symbols.

Given a list of formulas and a working set of input symbols (initially empty), in a loop, you:

1. Iterate through the list of formulas and find dependencies using the current working set as input/context
2. If there are no ambiguities and all types are known, you add the result type to the working set
3. You sort the failures by their degree of ambiguities, then go back to step 1

This process continues until no progress is made. If no progress is made in an iteration, then you return an error that the inference is intractable. Note the performance of such an algorithm might be very slow for large formula sets.